client/telemetry/src/lib.rs

```diff
@@ -16,339 +16,401 @@
 //!
+//! The macro [`telemetry`] can be used to report telemetries from anywhere in the code but the
+//! telemetry must have been initialized through [`TelemetryHandle::start_telemetry()`].
+//! Initializing the telemetry will make all the following code execution (including newly created
+//! async background tasks) report the telemetries through the endpoints provided during the
+//! initialization. If multiple telemetries have been started, the latest one (higher up in the
+//! stack) will be used. If no telemetry has been started, nothing will be reported.
+
+#![warn(missing_docs)]
 
-use futures::{prelude::*, channel::mpsc};
-use libp2p::{Multiaddr, wasm_ext};
+use futures::{channel::mpsc, prelude::*};
+use libp2p::{wasm_ext, Multiaddr};
 use log::{error, warn};
-use parking_lot::Mutex;
-use serde::{Serialize, Deserialize, Deserializer};
-use std::{pin::Pin, sync::Arc, task::{Context, Poll}, time::Duration};
+use sp_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver};
+use std::{
+	collections::{HashMap, HashSet},
+	time::Duration,
+};
+use tracing::Id;
 use wasm_timer::Instant;
 
+pub use chrono;
 pub use libp2p::wasm_ext::ExtTransport;
-pub use slog_scope::with_logger;
-pub use slog;
+pub use serde_json;
+pub use tracing;
+
+mod dispatcher;
+mod endpoints;
+mod layer;
+mod node;
+mod transport;
+
+use dispatcher::*;
+pub use endpoints::*;
+pub use layer::*;
+use node::*;
+use transport::*;
+
+/// Substrate DEBUG log level.
+pub const SUBSTRATE_DEBUG: u8 = 9;
+/// Substrate INFO log level.
+pub const SUBSTRATE_INFO: u8 = 0;
+
+/// Consensus TRACE log level.
+pub const CONSENSUS_TRACE: u8 = 9;
+/// Consensus DEBUG log level.
+pub const CONSENSUS_DEBUG: u8 = 5;
+/// Consensus WARN log level.
+pub const CONSENSUS_WARN: u8 = 4;
+/// Consensus INFO log level.
+pub const CONSENSUS_INFO: u8 = 1;
+
+pub(crate) type InitPayload = (
+	Id,
+	TelemetryEndpoints,
+	ConnectionMessage,
+	mpsc::UnboundedReceiver<ConnectionNotifierSender>,
+);
+
+/// Telemetry worker.
+///
+/// It should be ran as a background task using the [`TelemetryWorker::run`] method. This method
+/// will consume the object and any further attempts of initializing a new telemetry through its
+/// handle will fail (without being fatal).
+#[derive(Debug)]
+pub struct TelemetryWorker {
+	receiver: mpsc::Receiver<(Id, u8, String)>,
+	sender: mpsc::Sender<(Id, u8, String)>,
+	init_receiver: mpsc::UnboundedReceiver<InitPayload>,
+	init_sender: mpsc::UnboundedSender<InitPayload>,
+	transport: WsTrans,
+}
 
-mod async_record;
-mod worker;
+/// Telemetry Result typedef.
+pub type Result<T> = std::result::Result<T, Error>;
 
-/// Configuration for telemetry.
-pub struct TelemetryConfig {
-	/// Collection of telemetry WebSocket servers with a corresponding verbosity level.
-	pub endpoints: TelemetryEndpoints,
+/// Telemetry errors.
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+#[non_exhaustive]
+pub enum Error {
+	#[error(transparent)]
+	Io(#[from] std::io::Error),
```

Comment:
Why not just use `std::io::Error` as error?

Reply:
I have no idea... For what I can see in the commit I did it for sc-tracing and sc-telemetry. In sc-tracing it makes a lot more sense because there are many possible errors.

Maybe I wanted to make it clear it's an error during telemetry initialization/handling. That could make sense since IoError can come from anywhere. With this, it would be properly wrapped so we know this io error is related to telemetry.

:thinking: I'm dubious about this. What do you think? I should revert it to IoError?

Reply:
If we don't have more errors, yes. Requires less code.

Reply:
Ok so I was changing it as you said and I found out where it is used: https://github.com/paritytech/substrate/blob/cecton-telemetry-without-global/client/tracing/src/logging/mod.rs#L57

Have a look at this. You can see that if I use "IoError" there it means an IoError occurred but we have no way of knowing that it failed because of the telemetry.

So (sorry to ask again :pray:): are you 100% sure you think it's best to use IoError here instead of TelemetryError? Here is the change 0a8ca91d6